### PR TITLE
[move-compiler] Fix optimizations blocked by warnings

### PR DESCRIPTION
**Pull request by tnowacki**
- Fix compiler optimizations being erroneously blocked by all diagnostics, not just errors
- Fix internal ICE checks for the compiler's reference safety checks

## Motivation

- @awelc found a bug where optimizations weren't done in the presence of warnings.

## Test Plan

- Optimizations are not currently observable in the test suite

### CODE REVIEW

File: language/move-compiler/src/cfgir/borrows/mod.rs [hunk 1]
```diff
@@ -86,8 +86,9 @@ pub fn verify(
     locals: &UniqueMap<Var, SingleType>,
     cfg: &super::cfg::BlockCFG,
 ) -> BTreeMap<Label, BorrowState> {
-    let mut initial_state =
-        BorrowState::initial(locals, acquires.clone(), compilation_env.has_diags());
+    // check for existing errors
+    let has_errors = compilation_env.has_diags_at_or_above_severity(Severity::NonblockingError);
```
Comment (sblackshear) [inline language/move-compiler/src/cfgir/borrows/mod.rs:90]: Is a "warning" the same asa  "nonblocking error"?

File: language/move-compiler/src/command_line/compiler.rs [hunk 2]
```diff
@@ -465,7 +465,7 @@ pub fn construct_pre_compiled_lib<Paths: Into<Symbol>, NamedAddress: Into<Symbol
             typing = Some(tprog.clone())
         }
         PassResult::HLIR(hprog) => {
-            if env.has_diags() {
+            if env.has_diags_at_or_above_severity(Severity::NonblockingError) {
```
Comment (sblackshear) [inline language/move-compiler/src/command_line/compiler.rs:468]: The usage of `NoblockingError` here, but `BlockingError` above is puzzling to me.

**Comment by sblackshear** [comment]
One thought: there are a lot of usages of `has_diags` inside the compiler:

```
src/cfgir/borrows/mod.rs:        BorrowState::initial(locals, acquires.clone(), compilation_env.has_diags());
src/cfgir/translate.rs:            if !context.env.has_diags() {
src/command_line/compiler.rs:            if env.has_diags() {
src/command_line/compiler.rs:            if env.has_diags() {
src/command_line/compiler.rs:            if env.has_diags() {
src/expansion/byte_string.rs:    fn has_diags(&self) -> bool {
src/expansion/byte_string.rs:    if !context.has_diags() {
src/expansion/translate.rs:        assert!(context.env.has_diags())
src/expansion/translate.rs:        None => assert!(context.env.has_diags()),
src/expansion/translate.rs:        assert!(context.env.has_diags())
src/expansion/translate.rs:        assert!(context.env.has_diags())
src/expansion/translate.rs:            debug_assert!(context.env.has_diags());
src/expansion/translate.rs:                    assert!(context.env.has_diags());
src/expansion/translate.rs:                    assert!(context.env.has_diags());
src/expansion/translate.rs:                    assert!(context.env.has_diags());
src/expansion/translate.rs:                assert!(context.env.has_diags());
src/expansion/translate.rs:                    assert!(context.env.has_diags());
src/expansion/translate.rs:                    assert!(context.env.has_diags());
src/expansion/translate.rs:                    assert!(context.env.has_diags());
src/expansion/translate.rs:                        assert!(context.env.has_diags());
src/expansion/translate.rs:                        assert!(context.env.has_diags());
src/expansion/translate.rs:                    assert!(context.env.has_diags());
src/expansion/translate.rs:                assert!(context.env.has_diags());
src/hlir/translate.rs:        assert!(fields.is_some() || self.env.has_diags());
src/hlir/translate.rs:                        debug_assert!(context.env.has_diags());
src/hlir/translate.rs:                        debug_assert!(o.is_some() || context.env.has_diags());
src/hlir/translate.rs:            assert!(context.env.has_diags());
src/hlir/translate.rs:            assert!(context.env.has_diags());
src/naming/translate.rs:                    assert!(self.env.has_diags());
src/naming/translate.rs:                    assert!(self.env.has_diags());
src/naming/translate.rs:                    assert!(self.env.has_diags());
src/naming/translate.rs:        assert!(context.env.has_diags());
src/naming/translate.rs:            assert!(context.env.has_diags());
src/naming/translate.rs:                assert!(context.env.has_diags());
src/naming/translate.rs:                    assert!(context.env.has_diags());
src/naming/translate.rs:                    assert!(context.env.has_diags());
src/naming/translate.rs:                    assert!(context.env.has_diags());
src/naming/translate.rs:                    assert!(context.env.has_diags());
src/naming/translate.rs:                    assert!(context.env.has_diags());
src/naming/translate.rs:                    assert!(context.env.has_diags());
src/naming/translate.rs:                    assert!(context.env.has_diags());
src/naming/translate.rs:                assert!(context.env.has_diags());
src/naming/translate.rs:                            assert!(context.env.has_diags());
src/naming/translate.rs:                        assert!(context.env.has_diags());
src/naming/translate.rs:            assert!(context.env.has_diags());
src/naming/translate.rs:            assert!(context.env.has_diags());
src/shared/mod.rs:    pub fn has_diags(&self) -> bool {
src/typing/globals.rs:            assert!(context.env.has_diags());
src/typing/globals.rs:            assert!(context.env.has_diags());
src/typing/translate.rs:            assert!(context.env.has_diags());
```

But based on this PR, it seems like you (almost) always want to distinguish between "has errors" and "has warnings". Should we just get rid of "has_diags" in favor of those?

**Comment by tnowacki** [comment]
> But based on this PR, it seems like you (almost) always want to distinguish between "has errors" and "has warnings". Should we just get rid of "has_diags" in favor of those?

This is a very good suggestion. And simpler than what I am doing... so let me change that :P

**Review by sblackshear — APPROVED** [review]
Nice, I find this new version very readable!

**Review by awelc — APPROVED** [review]
LGTM

**Pull request merged**